packages/lib/src/verify.ts

```diff
@@ -1,199 +1,180 @@
-import { isNode } from "./utils";
+import {
+  BatchVerifyArgs,
+  VerificationResult,
+  VerifyArgs,
+  ZKP,
+  ZKPPublicSignals,
+} from "./types";
+import { areAllBigIntsDifferent, isNode } from "./utils";
 
 /**
  * Verifies an ECDSA membership proof
  * Performs in circuit and out of circuit verification
  * Based on the Efficient ECDSA formulation: https://personaelabs.org/posts/efficient-ecdsa-1/
- * Does not check/maintain the list of usedNullifiers, this must be done by the caller
+ * Does not maintain the list of usedSigNullifiers, this must be done by the caller
  * @param proof - The membership proof to verify
- * @param pubKeys - The list of public keys comprising the anonymity set for the proof
- * @param nullifierRandomness - Optional nullifier randomness used to generate unique nullifiers
+ * @param merkleRoot - Precomputed merkle root for the public key anonymity set
+ * @param merkleRootArgs - Arguments to generate the merkle root. Only needed if merkle root is not precomputed
+ * @param sigNullifierRandomness - Randomness used to generate signature nullifiers. Must be unique per application
+ * @param usedSigNullifiers - The list of used signature nullifiers. Used to prevent double proofs
  * @param pathToCircuits - The path to the verification key. Only needed for server side verification
- * @returns - A boolean indicating whether or not the proof is valid
+ * @param enableTiming - Whether or not to log timing information
+ * @returns - A boolean indicating whether or not the proof is valid, and a list of newly spent sig nullifiers
  */
-export const verifyMembership = async (
-  proof: EcdsaMembershipProof,
-  pubKeys: WeierstrassPoint[],
-  nullifierRandomness: bigint = BigInt(0),
-  pathToCircuits: string | undefined = undefined
-): Promise<boolean> => {
+export const verifyMembership = async ({
+  proof,
+  merkleRoot,
+  merkleRootArgs,
+  sigNullifierRandomness,
+  usedSigNullifiers,
+  pathToCircuits,
+  enableTiming,
+}: VerifyArgs): Promise<VerificationResult> => {
   if (isNode() && pathToCircuits === undefined) {
     throw new Error(
       "Path to circuits must be provided for server side verification!"
     );
   }
+  if (!merkleRoot && !merkleRootArgs) {
+    throw new Error("Must provide either merkle root or merkle root args!");
+  }
 
-  console.time("Membership Proof Verification");
-  console.time("Merkle Root Verification");
-  const publicSignals = proof.zkp.publicSignals;
-  const merkleRoot = BigInt(publicSignals[1]);
-  const edwardsPubKeys = pubKeys.map((pubKey) => pubKey.toEdwards());
-  const computedMerkleRoot = await computeMerkleRoot(edwardsPubKeys);
-  if (computedMerkleRoot !== merkleRoot) {
-    return false;
+  const timingUuid = uuidv4();
+  enableTiming && console.time(`Membership Proof Verification: ${timingUuid}`);
+  const publicSignals = getPublicSignalsFromMembershipZKP(proof.zkp);
+
+  enableTiming && console.time(`Merkle Root Verification: ${timingUuid}`);
```

Comment:
nit: should we do the same thing in the proving functions instead of the if? or keep the if everywher?

Reply:
Oops, meant to fix this, addressed.